### PR TITLE
Type hints

### PR DESCRIPTION
**Pull request by roycoding**
- Type hints for all source files
- Added mypy config file
- Added black config file

Reference #1

### CODE REVIEW

File: src/tsaug/affine.py [hunk 1]
```diff
@@ -2,13 +2,20 @@
...
 
 
 @dimensionalize
-def affine(X, Y=None, a=1.0, b=0.0):
+def affine(
+    X: np.ndarray,
+    Y: Optional[np.ndarray] = None,
+    a: Union[float, np.ndarray] = 1.0,
+    b: Union[float, np.ndarray] = 0.0,
+) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (tailaiw) [inline src/tsaug/affine.py]: The second output should be optional

File: src/tsaug/affine.py [hunk 2]
```diff
@@ -76,14 +88,14 @@ def affine(X, Y=None, a=1.0, b=0.0):
...
-    random_seed=None,
-):
+    X: np.ndarray,
+    Y: Optional[np.ndarray] = None,
+    max_a: float = 10.0,
+    min_a: float = -10.0,
+    max_b: float = 100.0,
+    min_b: float = -100.0,
+    random_seed: Optional[int] = None,
+) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (tailaiw) [inline src/tsaug/affine.py]: The second output should be optional

File: src/tsaug/cross_sum.py [hunk 3]
```diff
@@ -1,14 +1,19 @@
...
 from .augmentor import _Augmentor
 
 
 @dimensionalize
-def cross_sum(X, Y=None, inds=None):
+def cross_sum(
+    X: np.ndarray,
+    Y: Optional[np.ndarray] = None,
+    inds: Optional[np.ndarray] = None,
+) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (tailaiw) [inline src/tsaug/cross_sum.py]: The second output should be optional

File: src/tsaug/cross_sum.py [hunk 4]
```diff
@@ -68,19 +77,24 @@ def cross_sum(X, Y=None, inds=None):
...
 
 
 @dimensionalize
-def random_cross_sum(X, Y=None, max_sum_series=5, random_seed=None):
+def random_cross_sum(
+    X: np.ndarray,
+    Y: Optional[np.ndarray] = None,
+    max_sum_series: Optional[int] = 5,
+    random_seed: Optional[int] = None,
+) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (tailaiw) [inline src/tsaug/cross_sum.py]: The second output should be optional

File: src/tsaug/zoom.py [hunk 5]
```diff
@@ -11,14 +12,20 @@
...
+    size: Optional[int] = None,
+) -> Tuple[np.ndarray, Optional[np.ndarray]]:
     """Magnify time intervels of time series.
 
     This transformation does not change the number of time points in a series.
 
     Parameters
     ----------
-    X : numpy.ndarray
+    X : numpy.ndarray[sizes == size]
```
Comment (tailaiw) [inline src/tsaug/zoom.py]: I don't understand the syntax here

File: src/tsaug/zoom.py [hunk 6]
```diff
@@ -36,7 +43,7 @@ def magnify(X, Y=None, start=0, end=None, size=None):
 
     end : int or numpy.ndarray, optional
         Indices of the ending positions of time windows to be magnified. If
-        an integer, all series use the same value. Only used when argument
+        an integer, all series use the same value. Only used when argument[sizes == size]
```
Comment (tailaiw) [inline src/tsaug/zoom.py]: I don't understand the syntax here

File: src/tsaug/zoom.py [hunk 7]
```diff
@@ -50,51 +57,72 @@ def magnify(X, Y=None, start=0, end=None, size=None):
...
         if Y is None:
-            X_crop = crop(X, crop_start=start, crop_size=size)
+            X_crop = crop(
+                X, crop_start=start, crop_size=size
+            )  # type: np.ndarray
             return resample(X_crop, n_new=n)
         else:
+            Y_crop = (
+                []
+            )  # type: np.ndarray  # See earlier variable attribution hack
```
Comment (tailaiw) [inline src/tsaug/zoom.py:78]: What is the purpose of this line?

File: src/tsaug/dimensionalize.py [hunk 8]
```diff
@@ -1,21 +1,33 @@
...
     """
 
     @wraps(f)
-    def g(X, Y=None, *args, **kwargs):
+    def g(
+        X: np.ndarray,
+        Y: Optional[np.ndarray] = None,
+        *args: Any,
+        **kwargs: Any
+    ) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (tailaiw) [inline src/tsaug/dimensionalize.py]: The output should be a 2-tuple (numpy array, numpy array), or a numpy array

File: src/tsaug/augmentor.py [hunk 9]
```diff
@@ -1,15 +1,24 @@
...
+
 import numpy as np
 
 
 class _Augmentor:
-    def __init__(self, augmentor_func=None, is_random=None, **kwargs):
+    def __init__(
+        self,
+        augmentor_func: Optional[Callable[..., np.ndarray]] = None,
+        is_random: Optional[bool] = None,
```
Comment (tailaiw) [inline src/tsaug/augmentor.py]: This is a bug in my previous code. Argument `is_random` should be a required boolean.

File: src/tsaug/augmentor.py [hunk 10]
```diff
@@ -1,15 +1,24 @@
+from typing import List, Tuple, Union, Optional, Callable, Any
+
 import numpy as np
 
 
 class _Augmentor:
-    def __init__(self, augmentor_func=None, is_random=None, **kwargs):
+    def __init__(
+        self,
+        augmentor_func: Optional[Callable[..., np.ndarray]] = None,
```
Comment (tailaiw) [inline src/tsaug/augmentor.py]: This is a problem in my previous code. I think it makes more sense to set the argument `augmentor_func` as required.

Comment (tailaiw) [inline src/tsaug/augmentor.py]: I don't understand what the syntax `Callable[..., np.ndarray]]` means?

File: src/tsaug/augmentor.py [hunk 11]
```diff
@@ -1,15 +1,24 @@
...
+        **kwargs: Any
+    ) -> None:
         self._augmentor_func = augmentor_func
         self._is_random = is_random
-        self._M = 1
-        self._prob = 1.0
-        self._params = kwargs
+        self._M = 1  # type: int
+        self._prob = 1.0  # type: float
+        self._params = kwargs  # type: Any
```
Comment (tailaiw) [inline src/tsaug/augmentor.py]: `kwargs` is dict, right?

**Review by tailaiw — CHANGES_REQUESTED** [review]

File: src/tsaug/crop.py [hunk 12]
```diff
@@ -239,7 +279,12 @@ class RandomCrop(_Augmentor):
 
     """
 
-    def __init__(self, crop_size=None, crops_per_series=1, random_seed=None):
+    def __init__(
+        self,
+        crop_size: Optional[int] = None,
+        crops_per_series: Optional[int] = 1,
```
Comment (tailaiw) [inline src/tsaug/crop.py]: `crop_per_series` should be integer

File: src/tsaug/zoom.py (hunk 5 above)
Comment (roycoding) [reply src/tsaug/zoom.py]: It's possible that this was some sort of find and replace error while editing :thinking: 

I'm removing that.

File: src/tsaug/zoom.py (hunk 6 above)
Comment (roycoding) [reply src/tsaug/zoom.py]: Probably the same as the previous issue.

File: src/tsaug/zoom.py (hunk 7 above)
Comment (roycoding) [reply src/tsaug/zoom.py:78]: This is part of a horrible type hinting hack to handle the scenario where you have
1. Python 3.5
2. Multiple variables assigned in a single line (line 79 here)

File: src/tsaug/dimensionalize.py (hunk 8 above)
Comment (roycoding) [reply src/tsaug/dimensionalize.py]: Fixed

File: src/tsaug/augmentor.py (hunk 9 above)
Comment (roycoding) [reply src/tsaug/augmentor.py]: What should the default value be? @tailaiw

Comment (roycoding) [reply src/tsaug/augmentor.py]: Looking at the next comment, I think I understand now.

File: src/tsaug/augmentor.py (hunk 10 above)
Comment (roycoding) [reply src/tsaug/augmentor.py]: fixed

Comment (roycoding) [reply src/tsaug/augmentor.py]: The first item `...` means any number of arguments and the second is the return value, `np.ndarray`

File: src/tsaug/augmentor.py (hunk 11 above)
Comment (roycoding) [reply src/tsaug/augmentor.py]: yes, updated

File: src/tsaug/crop.py (hunk 12 above)
Comment (roycoding) [reply src/tsaug/crop.py]: fixed

**Review by tailaiw — APPROVED** [review]

**Pull request merged**